## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [2]:
# View the column names in the data
df.columns



Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [3]:
# Use the describe function to gather some basic statistics
df.describe()



client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [4]:
# Use this space to do any additional research
# and familiarize yourself with the data.



In [5]:
# What three item categories had the most entries?
df['category'].value_counts().head(3)



category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [7]:
# For the category with the most entries,
# which subcategory had the most entries?
df['subcategory'].value_counts().head(1)



subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [8]:
# Which five clients had the most entries in the data?
df['client_id'].value_counts().head(5)



client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [9]:
# Store the client ids of those top 5 clients in a list.
top_5_clients = df['client_id'].value_counts().head(5).index.tolist()
top_5_clients



[33615, 66037, 46820, 38378, 24741]

In [10]:
# How many total units (the qty column) did the
# client with the most entries order order?
df.loc[df['client_id'] == top_5_clients[0]]['qty'].sum()



64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [21]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty
df['subtotal'] = df['unit_price'] * df['qty']
df[['unit_price','qty','subtotal']].head()



unit_price  qty   subtotal
0     1096.80  105  115164.00
1       24.95   21     523.95
2       13.52   39     527.28
3       36.42   29    1056.18
4      195.10   20    3902.00

In [44]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.
def calc_shipping_price(x):
    if x > 50:
        return 7
    else:
        return 10

df['total_weight'] = df['unit_weight'] * df['qty']
df['shipping_price'] = df['total_weight'] * df['total_weight'].apply(calc_shipping_price)
df[['unit_price','unit_weight','qty','total_weight','shipping_price']].head()

unit_price  unit_weight  qty  total_weight  shipping_price
0     1096.80         7.50  105        787.50         5512.50
1       24.95         1.49   21         31.29          312.90
2       13.52         1.68   39         65.52          458.64
3       36.42         1.23   29         35.67          356.70
4      195.10        46.43   20        928.60         6500.20

In [47]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%
df['total_price'] = df['subtotal'] + df['shipping_price'] + ((df['subtotal'] + df['shipping_price']) * .0925).round(2)
df[['subtotal','shipping_price','total_price']].head()



subtotal  shipping_price  total_price
0  115164.00         5512.50    131839.08
1     523.95          312.90       914.26
2     527.28          458.64      1077.12
3    1056.18          356.70      1543.57
4    3902.00         6500.20     11364.40

In [48]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).
df['total_cost'] = df['unit_cost'] * df['qty'] + df['shipping_price']
df[['unit_cost','qty','shipping_price','total_cost']].head()



unit_cost  qty  shipping_price  total_cost
0     762.71  105         5512.50    85597.05
1      15.09   21          312.90      629.79
2       7.86   39          458.64      765.18
3      24.85   29          356.70     1077.35
4     108.17   20         6500.20     8663.60

In [49]:
# Create a column for the profit of
# each line using line cost and line price
df['profit'] = df['total_price'] - df['total_cost']
df[['total_price','total_cost','profit']].head()



total_price  total_cost    profit
0    131839.08    85597.05  46242.03
1       914.26      629.79    284.47
2      1077.12      765.18    311.94
3      1543.57     1077.35    466.22
4     11364.40     8663.60   2700.80

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [95]:
# Check your work using the totals above

orders_to_check = [2742071,2173913,6128929]

order_grouped_df = df.groupby('order_id').agg({'total_price':'sum'}).reset_index()

order_check_df = order_grouped_df[order_grouped_df['order_id'].isin(orders_to_check)]

for order in orders_to_check:
    print(f"Order {order} total: ${order_check_df.loc[order_check_df['order_id'] == order]['total_price'].values[0]:,.2f}")




Order 2742071 total: $152,811.89
Order 2173913 total: $162,388.71
Order 6128929 total: $923,441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [100]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.
client_grouped_df = df.groupby('client_id').agg({'total_price':'sum'}).reset_index()

for client in top_5_clients:
    print(f"Client {client} total: ${client_grouped_df.loc[client_grouped_df['client_id'] == client]['total_price'].values[0]:,.2f}")



Client 33615 total: $8,377,308.52
Client 66037 total: $10,259,514.79
Client 46820 total: $9,743,794.36
Client 38378 total: $12,906,550.87
Client 24741 total: $82,268,892.02


In [140]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.
def print_summary(df):
    for row in top_5_summary_df.iterrows():
        print(f"Client {row[1]['client_id']:.0f} : purchased {row[1]['qty']:,.0f} units, spent ${row[1]['shipping_price']:,.2f} on shipping, for a total price ${row[1]['total_price']:,.2f}. The Total Cost was ${row[1]['total_cost']:,.2f} resulting in {row[1]['profit']:,.2f} profit.")
    


summary_df = df.groupby('client_id').agg({'qty':'sum','shipping_price':'sum','total_price':'sum','total_cost':'sum','profit':'sum'}).reset_index()

top_5_summary_df = summary_df[summary_df['client_id'].isin(top_5_clients)].sort_values('profit',ascending=False).reset_index(drop=True)

print_summary(top_5_summary_df)




Client 24741 : purchased 239,862 units, spent $5,126,448.37 on shipping, for a total price $82,268,892.02. The Total Cost was $45,688,899.71 resulting in 36,579,992.31 profit.
Client 38378 : purchased 73,667 units, spent $3,429,455.40 on shipping, for a total price $12,906,550.87. The Total Cost was $9,634,720.98 resulting in 3,271,829.89 profit.
Client 66037 : purchased 43,018 units, spent $1,395,151.85 on shipping, for a total price $10,259,514.79. The Total Cost was $7,004,482.98 resulting in 3,255,031.81 profit.
Client 46820 : purchased 75,768 units, spent $1,601,448.84 on shipping, for a total price $9,743,794.36. The Total Cost was $7,007,191.64 resulting in 2,736,602.72 profit.
Client 33615 : purchased 64,313 units, spent $1,828,984.89 on shipping, for a total price $8,377,308.52. The Total Cost was $6,175,313.91 resulting in 2,201,994.61 profit.


In [18]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.



In [179]:
# Sort the updated data by "Total Profit" from highest to lowest
def convert_to_millions(x):
    return x / 1000000


sorted_top_5 = top_5_summary_df.sort_values('profit', ascending=False)

sorted_top_5 = sorted_top_5.rename(columns={'client_id': 'Client ID', 'qty': 'Units Purchased', 'shipping_price': 'Shipping Cost', 'total_price': 'Total Revenue', 'total_cost': 'Total Cost', 'profit': 'Total Profit'})

formatted_top_5 = sorted_top_5[['Client ID', 'Units Purchased', 'Shipping Cost', 'Total Revenue', 'Total Profit']]

#formatted_top_5.columns

formatted_top_5["Shipping Cost"] = formatted_top_5.loc[:,'Shipping Cost'].apply(convert_to_millions).map('${:,.2f}'.format) + 'M'
formatted_top_5['Total Revenue'] = formatted_top_5['Total Revenue'].apply(convert_to_millions).map('${:,.2f}'.format) + 'M'
formatted_top_5['Total Profit'] = formatted_top_5['Total Profit'].apply(convert_to_millions).map('${:,.2f}'.format) + 'M'

formatted_top_5


C:\Users\acrai\AppData\Local\Temp\ipykernel_15508\2889307379.py:14: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  formatted_top_5["Shipping Cost"] = formatted_top_5.loc[:,'Shipping Cost'].apply(convert_to_millions).map('${:,.2f}'.format) + 'M'
C:\Users\acrai\AppData\Local\Temp\ipykernel_15508\2889307379.py:15: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  formatted_top_5['Total Revenue'] = formatted_top_5['Total Revenue'].apply(convert_to_millions).map('${:,.2f}'.format) + 'M'
C:\Users\acrai\AppData\Local\T

Client ID  Units Purchased Shipping Cost Total Revenue Total Profit
0      24741           239862        $5.13M       $82.27M      $36.58M
1      38378            73667        $3.43M       $12.91M       $3.27M
2      66037            43018        $1.40M       $10.26M       $3.26M
3      46820            75768        $1.60M        $9.74M       $2.74M
4      33615            64313        $1.83M        $8.38M       $2.20M